atlas-spec/mpc-engine/src/party.rs

```diff
@@ -76,144 +207,284 @@ impl Party {
+            .send(SubMessage::OTCommit(sender_commitment))
+            .unwrap();
+
+        let selection_msg = my_inbox.recv().unwrap();
+        if let SubMessage::OTSelect(selection) = selection_msg {
+            let payload =
+                sender_state.send(left_input, right_input, &selection, &mut self.entropy)?;
+
+            receiver_address.send(SubMessage::OTSend(payload)).unwrap();
             Ok(())
         } else {
-            Err(Error::UnexpectedSubprotocolMessage(receiver_msg))
+            Err(Error::UnexpectedSubprotocolMessage(selection_msg))
         }
     }
 
     /// Listen for an OT initiation as the receiver.
     ///
-    /// The receiver needs to provide a choice of left or right output to the
-    /// protocol and receives the chosen sender input.
-    fn ot_receive(&mut self, choose_left: bool) -> Result<Vec<u8>, Error> {
-        let channel_msg = self.channels.listen.recv().unwrap();
-
-        if let Message {
-            to,
-            from,
-            payload: MessagePayload::SubChannel(their_channel, my_channel),
-        } = channel_msg
-        {
-            let first_msg = my_channel.recv().unwrap();
-            if let SubMessage::OTCommit(commitment) = first_msg {
-                let dst = format!("OT-{}-{}", from, to);
-                self.log(&format!("Choose left input: {choose_left}"));
-                let (receiver, resp) = crate::primitives::ot::OTReceiver::select(
-                    &mut self.entropy,
-                    dst.as_bytes(),
-                    commitment,
-                    choose_left,
-                )?;
-                their_channel.send(SubMessage::OTSelect(resp)).unwrap();
-                let second_msg = my_channel.recv().unwrap();
-                if let SubMessage::OTSend(payload) = second_msg {
-                    assert_eq!(self.id, to);
-                    let result = receiver.receive(payload)?;
-                    self.log(&format!("Got message {result:?}"));
-                    Ok(result)
-                } else {
-                    Err(Error::UnexpectedSubprotocolMessage(second_msg))
-                }
+    /// The receiver needs to provide a choice of left or right sender input to
+    /// the protocol and receives the chosen sender input.
+    fn ot_receive(
+        &mut self,
+        choose_left_input: bool,
+        sender_address: Sender<SubMessage>,
+        my_inbox: Receiver<SubMessage>,
+        sender_id: usize,
+    ) -> Result<Vec<u8>, Error> {
+        let ot_commit_msg = my_inbox.recv().unwrap();
+        if let SubMessage::OTCommit(commitment) = ot_commit_msg {
+            let domain_separator = format!("OT-{}-{}", sender_id, self.id);
+            let (receiver_state, receiver_selection) = crate::primitives::ot::OTReceiver::select(
+                &mut self.entropy,
+                domain_separator.as_bytes(),
+                commitment,
+                choose_left_input,
+            )?;
+            sender_address
```

Comment:
What's your policy on when to use errors when to unwrap? Using both is fine, esp. in a spec. But I don't think I understand when you use what. Maybe out some description for this into the comments for errors or the library in general.

Reply:
That's a very good point! I would actually err on the side of unwrapping, especially with insufficient randomness since that is not recoverable. There is some subtlety there with underlying crypto errors, since some of them could indicate a party trying to cheat and that information should be passed on to whatever application is using the spec in order to gracefully (and securely) deal with aborting the protocol or excluding the cheater. The channel `Results` I pretty much always want to unwrap, since an `Error` here indicates a bug in the specification (should `expect`, really then).
I should be consistent and document all that though. :+1:

Reply:
Added some documentation describing this policy and made sure we `panic!` if we have insufficient randomness. Going forward, I'll take care to review when an error might indicate cheating (in which case it should be passed on) or just a bug (in which case I would be fine `panic!`ing).